### ISSUE TITLE
ICE in clippy_lints::matches::Matches

### ISSUE THREAD

**Issue by digama0** [rust-lang/rust-clippy#7012]
### Code

[playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4e3e2945699bd54ee82aeaecd927fd38)
```rust
#![allow(clippy::single_match)]

enum _MyOption {
  None,
  Some(()),
}

impl _MyOption {
  fn _foo(&self) {
    match self {
      &Self::Some(_) => {},
      _ => {}
    }
  }
}

fn main() {}
```

### Meta

- `cargo clippy -V`: clippy 0.1.53 (74874a6 2021-03-30)
- `rustc -Vv`:
  ```
  rustc 1.53.0-nightly (74874a690 2021-03-30)
  binary: rustc
  commit-hash: 74874a690bc95443292496ff5df5cc5c8cb56e0b
  commit-date: 2021-03-30
  host: x86_64-unknown-linux-gnu
  release: 1.53.0-nightly
  LLVM version: 12.0.0
  ```

### Error output

```
$ RUST_BACKTRACE=full cargo clippy
    Checking rust-test v0.1.0 (/home/mario/Documents/rust-test)
thread 'rustc' panicked at 'attempted .def_id() on invalid res: Err', /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/compiler/rustc_hir/src/def.rs:409:32
stack backtrace:
   0:     0x7f9574ff3d90 - std::backtrace_rs::backtrace::libunwind::trace::h4dee703919bfd40a
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x7f9574ff3d90 - std::backtrace_rs::backtrace::trace_unsynchronized::h457e839f1a563e20
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f9574ff3d90 - std::sys_common::backtrace::_print_fmt::h86a55fb30f8393c8
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f9574ff3d90 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h7b3d6cac46d277e1
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7f9575064f9f - core::fmt::write::h127419eb46f2ecc9
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/core/src/fmt/mod.rs:1094:17
   5:     0x7f9574fe8855 - std::io::Write::write_fmt::h6010cfbb4726588b
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/io/mod.rs:1578:15
   6:     0x7f9574ff798b - std::sys_common::backtrace::_print::h79b4f9652330cc9d
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7f9574ff798b - std::sys_common::backtrace::print::h330bb326a76af8cf
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7f9574ff798b - std::panicking::default_hook::{{closure}}::heb6a42a7d50a472e
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/panicking.rs:208:50
   9:     0x7f9574ff746d - std::panicking::default_hook::h17e521ba6d68d6e1
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/panicking.rs:225:9
  10:     0x55e008eb0279 - clippy_driver::ICE_HOOK::{{closure}}::{{closure}}::ha2d5d9757a08a932
  11:     0x7f9574ff80a0 - std::panicking::rust_panic_with_hook::h70db735e3a6e70cb
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/panicking.rs:595:17
  12:     0x7f9574ff7c77 - std::panicking::begin_panic_handler::{{closure}}::h777c71c8e5a7e25c
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/panicking.rs:497:13
  13:     0x7f9574ff424c - std::sys_common::backtrace::__rust_end_short_backtrace::h3e9bf30168899554
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/sys_common/backtrace.rs:141:18
  14:     0x7f9574ff7bd9 - rust_begin_unwind
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/panicking.rs:493:5
  15:     0x7f9574fc3d7b - std::panicking::begin_panic_fmt::h0b6b6d9d830486ef
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/panicking.rs:435:5
  16:     0x55e008f79ea8 - rustc_hir::def::Res<Id>::def_id::{{closure}}::h96ee16f8264c1e05
  17:     0x55e008f951f5 - clippy_utils::recurse_or_patterns::h01bd70659a4d4fa0
  18:     0x55e009081fe7 - <clippy_lints::matches::Matches as rustc_lint::passes::LateLintPass>::check_expr::h1ef6e4539945496c
  19:     0x7f9576446834 - <rustc_lint::late::LateLintPassObjects as rustc_lint::passes::LateLintPass>::check_expr::h73ba55cd801c9cb2
  20:     0x7f95758b94ec - <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_expr::he0d9c1ca3660ce9f
  21:     0x7f95758b9669 - <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_block::h6c387cb4690a0702
  22:     0x7f95758b94f7 - <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_expr::he0d9c1ca3660ce9f
  23:     0x7f95758b9b0b - <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_nested_body::h7364eb428db9d346
  24:     0x7f95758b9d4a - <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_fn::h62e607c27bd5694f
  25:     0x7f95758b6001 - rustc_hir::intravisit::walk_impl_item::h1ad5990d6a8d426a
  26:     0x7f95758ba8c4 - rustc_hir::intravisit::Visitor::visit_nested_impl_item::h2646f7ecc46aca76
  27:     0x7f95758b6531 - rustc_hir::intravisit::walk_impl_item_ref::hac5c66b23ce7cd03
  28:     0x7f95758b75e5 - rustc_hir::intravisit::walk_item::hecb54583b013874f
  29:     0x7f95758ba2d0 - rustc_hir::intravisit::Visitor::visit_nested_item::h092aab6dba9f5c7e
  30:     0x7f95758b6acc - rustc_hir::intravisit::walk_mod::h75c5e3e8c8399984
  31:     0x7f95758b9e7f - <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_mod::ha7b4cf1f0d0d7be6
  32:     0x7f95758b5972 - rustc_hir::intravisit::walk_crate::he50b933f20b46f61
  33:     0x7f95758b7963 - rustc_lint::late::late_lint_pass_crate::heca5cdae055731c6
  34:     0x7f9577243673 - rustc_lint::late::late_lint_crate::h28cb0c90d3648839
  35:     0x7f957721755e - rustc_data_structures::sync::join::h13457a9e13b81256
  36:     0x7f957722dded - std::panic::catch_unwind::hcc5b94e09e686714
  37:     0x7f95772170ff - rustc_session::utils::<impl rustc_session::session::Session>::time::hf07fee08f48a4142
  38:     0x7f95772225ec - rustc_interface::passes::analysis::h75d132c9800eb84d
  39:     0x7f9575e62baa - rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps::h5c9d6ec206ec4bbe
  40:     0x7f95776906d0 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::hb30a4fbfea79270e
  41:     0x7f95776e3a67 - rustc_data_structures::stack::ensure_sufficient_stack::h3e61040cd98d1ba5
  42:     0x7f95775d1281 - rustc_query_system::query::plumbing::force_query_with_job::hcf939d362f9425e6
  43:     0x7f9577586503 - rustc_query_system::query::plumbing::get_query_impl::h7133bdb8ee936bd9
  44:     0x7f957766985f - <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis::h3b6d31f1707185c6
  45:     0x7f9577211657 - rustc_interface::passes::QueryContext::enter::h8ea88c76fc5b8911
  46:     0x7f95771e4d55 - rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter::he4dda08baf61c2dd
  47:     0x7f95771e3cd4 - rustc_span::with_source_map::hd041adcdcac51a55
  48:     0x7f95771e5c71 - rustc_interface::interface::create_compiler_and_run::hf35468725bf28979
  49:     0x7f95771ff288 - scoped_tls::ScopedKey<T>::set::h839b61bd5f63fdcc
  50:     0x7f95771ff7a4 - std::sys_common::backtrace::__rust_begin_short_backtrace::hbee7b442d38b6a0d
  51:     0x7f9577202ad5 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h6f6381cee62ab244
  52:     0x7f9575007f43 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7f7bdfc757a1add0
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/alloc/src/boxed.rs:1546:9
  53:     0x7f9575007f43 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha294ad0fb99e9b3a
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/alloc/src/boxed.rs:1546:9
  54:     0x7f9575007f43 - std::sys::unix::thread::Thread::new::thread_start::h4614854ae73cd6b0
                               at /rustc/74874a690bc95443292496ff5df5cc5c8cb56e0b/library/std/src/sys/unix/thread.rs:71:17
  55:     0x7f9574f05609 - start_thread
  56:     0x7f9574cdb293 - clone
  57:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.53 (74874a6 2021-03-30)

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `rust-test`

To learn more, run the command again with --verbose.
```

**Comment by digama0** [comment]
The error appears to be here:

https://github.com/rust-lang/rust-clippy/blob/8cf7d9b037bbdf5cd970b2f133ddd7d39acc1588/clippy_lints/src/matches.rs#L1044-L1047

The call to `qpath_res()` fails, so `def_id()` causes an ICE. The fact that the pattern match uses `Self::Variant` to refer to the variants seems to be related. This is also a recent breakage (within the last week), although I haven't bisected clippy history yet.

**Comment by digama0** [comment]
This is broken on `nightly-2021-03-27` = `clippy 0.1.53 (5e65467 2021-03-26)` , and works on `nightly-2021-03-26` = `clippy 0.1.52 (52e3dff 2021-03-25)`.
